Timeline UI/UX Improvements as well as new actions

CandyFace opened the pull request:
This PR invalidates #1287 by finishing the work where @4v0v left off, enhancing the features written and with added support for all layer types.

### New Description:
This PR introduces various timeline enhancements.
+ Adding frame next to current position, no matter if there's a frame or not.
  - Happens out of the box, just place the scrubber and add a frame. 
+ Add/Remove exposure:
  - Requirement: Frame has to be selected
  - How to use:
    - Shortcuts: Ctrl+'+' for add and Ctrl+'-' for subtract.
    - Menu: Animation -> 'Timeline selection' -> Add exposure / Subtract exposure
+ Add keyframe by double-clicking on the timeline.
  - The frame graphic while hovering
+ Scrubber will be shown as hollow when hovering
+ Ability to copy/cut/paste frames to timeline via standard shortcuts 
+ Ability to remove a selection of frames
  - Shortcuts: None as default but it can be set. 
  - Cannot be undone currently! Awaiting the new undo/redo implementation.
+ Reverse selection of frames.
  - Shortcuts: None as default but can be set.


### Old outdated description
~~
This allows you to:
* Add an empty frame at current position + 1
  - default shortcut has been set to "-" 
* Remove empty frame :
  - default shortcut has been set to "ctrl + -" 
* Insert a keyframe at current position + 1 and navigate to it
  - No new action for this, this logic replaces the old add keframe behavior.
* Copy/Cut an arbitrary selection of frames from timeline into clipboard.
* Paste selection to timeline
* Remove selection of frames
  - No shortcut currently
* Reverse selection of frames
  - Shortcut can be set but no default.

Additionally this also fixes that copy/cut and paste were always enabled, even though they did nothing.

**Works on all layer types.**

Tweaks:
* Only drawing tools will create a new frame (eg. not eraser)

It's practically ready to go although I would like some comments as to where we should put the "Remove selected frames" and "Reverse selected frames" as I feel they don't fit in "Animation".
~~

closes https://github.com/pencil2d/pencil/issues/575

CandyFace commented:
I've made a few QoL improvements
- When removing a keyframe, the scrubber won't move unless there's no keyframe
- The old add keyframe logic has been replaced with the insert insert logic, this should be preferred as it's more common case to want to add something next to x instead of next to an available spot. I'm more for simplicity here rather than options, so there's a good scenario which the old logic, I think we should use the new one as a replacement.
- Removed duplicate menu points.

CandyFace commented:
Given that I've added the ability to add keyframes via mouse interaction, I thought it would be ideal to also show where the frame will be placed before it happens. So here's another proposal:

While hovering over frames on the current layer track, the frame container will be "ghosted" so you will be able to see where the frame will be placed. 
Additionally, you can move to a frame by double-clicking, and double-clicking again on the current frame, will place a new frame container.

The fact that you have to be on the frame you want to add before you can place another frame, makes the action more explicit and the action also becomes more visual, because you can see the selection switching between the new frame and the previous frame.

Here's how it works:
![mouse-ux](https://user-images.githubusercontent.com/1045397/93711442-a3a8f000-fb4e-11ea-847d-a440801ebd01.gif)

If you want to try it without compiling the build yourself, I have a windows build ready.
https://drive.google.com/file/d/1-XDgagE9x7Jdq-iTTNT_OC0r6eNtIboc/view?usp=sharing

J5lx reviewed (changes requested):
I like where this PR is going, but unfortunately it still needs some work. First of all, I’ve submitted a PR to your repo with some cleanup etc (CandyFace#11), no need to clutter the review with it that way. As far as the other issues go, some are in the inline comments, but most of the major stuff is listed below:

- [ ] First of all, about your question from the OP regarding the menu entries. Personally I think the remove/reverse selected frames actions would best fit in the edit menu. However, if we move them there, some of the other entries currently in the animation menu should probably go there as well IMO, for example add/remove/duplicate frame

#### Problems

- [x] Copying/cutting from canvas seems always disabled on vector layers
- [x] Copying/cutting from canvas segfaults on bitmap layers
- [x] I haven’t been able to copy image data from other programs into Pencil2D (works in 0.6.5)
- [x] (I haven’t been able to copy image data from Pencil2D to other programs because of the segfault)
- [x] The new shortcuts are still missing in ShortcutsPage
- [ ] Reversing selected frames doesn’t immediately update the frames as displayed on the canvas, the changes are only reflected on the canvas after other interactions such as zooming. I think there was some talk about this but I don’t remember the details.
- [ ] Inserting a new keyframe while on an existing keyframe that is part of a group of consecutive keys but not the last one of them, it initially seems like a duplicated keyframe is appended at the end instead of inserting an empty one near the current position. This is also fixed by zooming.
- [x] Decreasing frame exposure moves not only the following key, but also any other selected keys
- [x] In 0.6.5, using copy when there’s no selection copies the entire frame content. With this PR, that is no longer possible.

#### Usability stuff

Just some things I noticed, I suppose not all of these are necessarily deal breakers.

- [x] I like the “preview” when hovering over the timeline, but I find it a little confusing that you have to double-click twice to create a new frame. I believe Jose mentioned this as well. I get why it’s done like that, but it’s confusing nonetheless. Perhaps it would be better to remove the double-click-to-scrub functionality altogether and leave scrubbing entirely to the timeline header bar. Then, double-click could always be used for inserting a frame. Just an idea.
- [x] “Remove selection of frames” should have shortcut support even if there’s no default
- [ ] I really appreciate that you tried to keep a single copy/cut/paste action, but at least in its current form it’s a little counterintuitive. Imagine making a selection of frames, then forgetting about it. Later you make a selection on canvas and try to copy it, but it doesn’t do what you expected/intended. This could be especially frustrating when trying to copy from Pencil2D to other programs, where you don’t see a change in the timeline once you try to paste.
  - [x] Related: Copy/Cut/Paste works on both canvas and timeline, but (de)select all only works on the canvas
- [x] The current frame selection is lost when increasing or decreasing frame exposure

#### Behaviour inconsistencies

Again just some things I noticed, not necessarily a big deal.

- [x] Inserting a new keyframe on an existing one followed by another keyframe clears the frame selection, but inserting a new keyframe on or right before an empty frame does not
- [x] ~~Increasing frame exposure moves the next block of consecutive keys, but decreasing frame exposure moves only the single following key~~ See https://github.com/pencil2d/pencil/pull/1343#issuecomment-741589307
- [x] Pasting on canvas replaces the contents of an existing selection, but pasting on the timeline ignores any existing selection

That’s quite a lot of remarks of course, and to be honest I feel like I’m kinda losing track of the big picture myself. Maybe it’s just because I’ve been working on this review slowly over such a long period of time, so I hope it’s not too overwhelming for you. It’s probably gonna be easier if you focus just on the inline comments and the “problems” section above at first. Thanks for all the work so far!

CandyFace commented:
Thanks for the review, I'll look into the issues you mentioned.

As for the usability points:
>  I like the “preview” when hovering over the timeline, but I find it a little confusing that you have to double-click twice to create a new frame. I believe Jose mentioned this as well. I get why it’s done like that, but it’s confusing nonetheless. Perhaps it would be better to remove the double-click-to-scrub functionality altogether and leave scrubbing entirely to the timeline header bar. Then, double-click could always be used for inserting a frame. Just an idea.

I think the solution to that could instead be to always create a keyframe a scrub to it when you double-click. I agree that double-clicking to scrub and then double-click again to create a keyframe can be pretty confusing The problem with double-clicking only creating a keyframe and not scrubbing is that it doesn't quite make sense if you ask me. You create a frame because you presumably are going to work on that frame, so if you have to make a second click somewhere else to select that created frame, then the action becomes needlessly complicated.

The proposed solution works like this:
![scrubber](https://user-images.githubusercontent.com/1045397/101597070-773a9c80-39f6-11eb-8ff6-1201421c67be.gif)

>  I really appreciate that you tried to keep a single copy/cut/paste action, but at least in its current form it’s a little counterintuitive. Imagine making a selection of frames, then forgetting about it. Later you make a selection on canvas and try to copy it, but it doesn’t do what you expected/intended. This could be especially frustrating when trying to copy from Pencil2D to other programs, where you don’t see a change in the timeline once you try to paste. 

Right.. I went the wrong way about that, you're right.. the clipboard should be kept in a context such that when you copy from the timeline, it will be kept there until either clear that context or make a new copy, the clipboard on canvas should not keep the same context as the timeline.
I will look into fixing it as well as the other copy/paste problems you mentioned, though I don't get a segfault, so that will be interesting to test... 

>  Increasing frame exposure moves the next block of consecutive keys, but decreasing frame exposure moves only the single following key

I will into the rest ;)

That's intended, and my reasoning for doing it like this is that you're either pushing or pulling frames (think physical blocks), you push a frame and only when it hits another frame, the next frame is pushed too, the same does not apply if you pull a frame, because you only pull one frame (they're not glued together :P). The behaviour is not consistent because the actions they perform are different, not necessary opposites of each other. If the behaviour should be consistent, then how would you go about decrease the exposure of only one frame, given that the frame is next to each other?

J5lx commented:
> I think the solution to that could instead be to always create a keyframe a scrub to it when you double-click.

That sounds like a much more natural solution actually! I should have thought of that myself…

> the clipboard should be kept in a context

I’m not sure I understand what you have in mind yet. Do you essentially mean maintaining two separate clipboards for the canvas and the timeline? In that case we would still need to find a way to determine what the user wants to do / which clipboard “context” they intend to act on. I imagine it might also be a little counterintuitive since normally there’s only one global clipboard.

And secondly, there’s still that issue with the way the clipboard is currently implemented in this PR (the one I posted in the inline comments). I imagine keeping separate clipboards would only work by keeping that sort of weird hybrid clipboard which this PR currently implements, and that would mean sacrificing the ability to copy/paste from/to other applications, or even just between separate Pencil2D processes.

> I don't get a segfault, so that will be interesting to test...

Interesting indeed… Please lmk when you need me to have a closer look at it once you start working on that.

> That's intended

Fair enough, I’ll strike it off

CandyFace commented on app/ui/mainwindow2.ui:
```diff
@@ -625,6 +631,28 @@
+    <string>Increase Frame Exposure</string>
+   </property>
+   <property name="toolTip">
+    <string>Adds empty frame next to current frame, will push all frames in front of current frame</string>
```
Neither seems to do anything on mac for menu items, will remove

CandyFace commented on core_lib/src/structure/layer.cpp:
```diff
@@ -207,12 +215,34 @@ bool Layer::removeKeyFrame(int position)
 
+void Layer::removeFromSelectionList(int pos)
+{
+    int count = 0;
+    for (int selectedPos : mSelectedFrames_byLast) {
+        if (selectedPos == pos) {
+            mSelectedFrames_byLast.removeAt(count);
+        }
+        count++;
+    }
+
+    count = 0;
+    for (int selectedPos : mSelectedFrames_byPosition) {
+        if (selectedPos == pos) {
+            mSelectedFrames_byPosition.removeAt(count);
+        }
+        count++;
+    }
```
I didn't know that one, handy indeed. will change.

CandyFace commented:
> Copying/cutting from canvas seems always disabled on vector layers
Fixed
> Copying/cutting from canvas segfaults on bitmap layers
Can't reproduce, check if it still happens now.

> I haven’t been able to copy image data from other programs into Pencil2D (works in 0.6.5)
Fixed
> (I haven’t been able to copy image data from Pencil2D to other programs because of the segfault)
Fixed
> The new shortcuts are still missing in ShortcutsPage
Fixed
> Reversing selected frames doesn’t immediately update the frames as displayed on the canvas, the changes are only reflected on the canvas after other interactions such as zooming. I think there was some talk about this but I don’t remember the details.
Fixed

> Inserting a new keyframe while on an existing keyframe that is part of a group of consecutive keys but not the last one of them, it initially seems like a duplicated keyframe is appended at the end instead of inserting an empty one near the current position. This is also fixed by zooming.
Sounds like the bug that should be addressed via https://github.com/pencil2d/pencil/pull/1497

> Decreasing frame exposure moves not only the following key, but also any other selected keys
Fixed.

> In 0.6.5, using copy when there’s no selection copies the entire frame content. With this PR, that is no longer possible.
Fixed, although the behaviour will still be a bit weird when there is both a selection on the timeline and on the canvas. Currently the behaviour will be that the canvas has priority over the timeline. If this turns out to be very confusing, then we can rethink the feature but then we probably will not be able to make it work with the standard ctrl+x, ctrl+c, ctrl+v shortcuts...

________________________________________________________________
> I like the “preview” when hovering over the timeline, but I find it a little confusing that you have to double-click twice to create a new frame. I believe Jose mentioned this as well. I get why it’s done like that, but it’s confusing nonetheless. Perhaps it would be better to remove the double-click-to-scrub functionality altogether and leave scrubbing entirely to the timeline header bar. Then, double-click could always be used for inserting a frame. Just an idea.
Fixed by the proposal mentioned earlier

> “Remove selection of frames” should have shortcut support even if there’s no default
Implemented

> I really appreciate that you tried to keep a single copy/cut/paste action, but at least in its current form it’s a little counterintuitive. Imagine making a selection of frames, then forgetting about it. Later you make a selection on canvas and try to copy it, but it doesn’t do what you expected/intended. This could be especially frustrating when trying to copy from Pencil2D to other programs, where you don’t see a change in the timeline once you try to paste.

The selection logic in pencil2D is terrible and very non intuitive as it is, it doesn't follow any conventional rules so there's definitely room for improvement here. I've made it so that canvas selections are prioritized now but there may still be situations that could be confusing if you have multiple selected frames and hope to get a copy of the canvas. This wouldn't be big deal normally, because you can always undo... except we can't because our we're still waiting for the undo/redo implementation I proposed forever ago.

> Related: Copy/Cut/Paste works on both canvas and timeline, but (de)select all only works on the canvas
Fixed although this might not be what we want... there doesn't have to be consistency between two different interfaces. 

> The current frame selection is lost when increasing or decreasing frame exposure
Fixed... but maybe not the best way to go about it.

I think i've addressed all remarks now in one way or another.

J5lx reviewed (changes requested):
Here we go again! I struggled to get anything done lately, so it took a while, sorry for the wait! Overall the PR is in a pretty good shape at this point, there are only a few issues left. First of all, I made another PR with all the minor, mostly cosmetic stuff: CandyFace#12. As for the “actual” issues:

- After selecting a few frames on the timeline and then switching to another layer, click-dragging over the previously selected frames switches back to the original layer and restores the selection. However, when switching back to the original layer using the layer list, the selection is not restored (as one would expect).
- Create a group of adjacent keyframes, then select the last keyframe and increase the exposure of the first one. When you do that, the keyframes after the first one will be moved back, but the selection will stay in place – that is, afterwards the keyframe preceding the previously selected keyframe will become the new selected keyframe. Once the selection has “moved” to the beginning of the group in this way, extending the exposure any further will simply cause the selected frame to be unselected. Maybe this is what you meant in your previous comment?
- After pasting onto the canvas, strokes made with the pencil tool will not appear within the selected area until the stroke is finished. The area outside the selection is not affected. There is no such issue in 0.6.5 and selections made manually with the select tool are also unaffected. Modifying the selection with the select tool after pasting also causes the issue to disappear.
- The frame reversal issue isn’t quite gone, but after merging with master everything is fine, so I suppose the PR you mentioned does indeed do the trick. Everything okay there.
- I get what you’re going for with ClipboardManager, but as it is now the way some of the logic is split between Editor and ClipboardManager is a little suboptimal. Especially for the paste part, where the *ClipboardManager* essentially has the knowledge of when *Editor* is able to perform a paste. I’d prefer if the separation was a bit cleaner there, for example by managing only the clipboard contents in ClipboardManager and keeping the copying/pasting entirely within Editor. But tbh, it doesn’t matter much where exactly you draw the line between ClipboardManager and Editor as long as it is a *clean* line.
- I like the “unified copy/paste” for canvas and timeline behaviour a little better than before, but it’s still not particularly elegant. I considered that maybe it would help to keep track of which selection (timeline vs canvas) was most recently modified and pick one according to that, but I’m worried that it might make the entire mechanism seem somewhat unpredictable. I’d like to hear what @Jose-Moreno thinks about this matter.

Thanks for addressing all those points from my last review!

CandyFace commented on core_lib/data/resources/kb.ini:
```diff
@@ -54,6 +54,10 @@ CmdMoveFrameForward=Ctrl+.
 CmdMoveFrameBackward=ctrl+","
 CmdAddFrame=F7
 CmdDuplicateFrame=F6
+CmdIncreaseFrameExposure=-
```
We can change it to something else then, I have no stance really. I didn't think long and hard, I just picked one that wasn't bound already because It the shortcut was never my focus.

J5lx replied:
Alright. I’d say let’s go with + for increase and - for decrease then, easy to remember.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+        if (clipboardImage.isNull()) { return false; }
+
+        // Clipboard is upt to date
+        if (*mBitmapImage.image() == clipboardImage) {
```
I didn't consider looking at the documentation when I implemented this, I thought the comparison would be more naive but I see it's doing a pixel by pixel search, so it's probably right to assume that there isn't much benefit in checking first.

CandyFace commented on app/src/mainwindow2.cpp:
```diff
@@ -1415,16 +1430,22 @@ void MainWindow2::makeConnections(Editor* pEditor, TimeLine* pTimeline)
 
+    // Clipboard state handling
+    connect(QApplication::clipboard(), &QClipboard::dataChanged, this, &MainWindow2::fetchClipboard);
+    connect(ui->menuEdit, &QMenu::aboutToShow, this, &MainWindow2::updateCopyCutPasteEnabled);
```
Yes, it ensures that the copy/cut/paste menu action will never show a false state, where as the other connections does not necessarily cover all cases. Don't remember which anymore...

CandyFace commented on core_lib/src/interface/editor.cpp:
```diff
@@ -508,108 +502,193 @@ void Editor::resetAutoSaveCounter()
+
+    if (!canCopy) { return; }
+
+    backup("Copy");
```
The action should be possible perform to trigger the menu highlight, so that the user can see that they activated the shortcut, otherwise they might become confused or infuriating at the keyboard for seemingly not doing what the user asks of it. Whether anything happens though, that's a different matter if you ask me.

I will fix the string

CandyFace commented on core_lib/src/managers/clipboardmanager.h:
```diff
@@ -0,0 +1,72 @@
+class QClipboard;
+class KeyFrame;
+
+class ClipboardManager: QObject
```
Fair enough, I didn't make use of the boilerplate methods, so it didn't make sense at the time but I agree it makes sense from a consistency perspective.

CandyFace commented on core_lib/src/interface/editor.cpp:
```diff
@@ -508,108 +502,193 @@ void Editor::resetAutoSaveCounter()
+
+    if (clipboards()->getClipboardFrames().empty()) {
+
+        if (clipboards()->canPaste(currentLayer)) {
```
Good point.. I must have missed something or made a mistake while reworking something here... it should be an early return if it's not possible to paste and otherwise backup if it's not a frame selection, since we can't undo a timeline frame selection.

CandyFace commented:
> After pasting onto the canvas, strokes made with the pencil tool will not appear within the selected area until the stroke is finished. The area outside the selection is not affected. There is no such issue in 0.6.5 and selections made manually with the select tool are also unaffected. Modifying the selection with the select tool after pasting also causes the issue to disappear.

Are you certain this issue doesn't exist in the nightly builds? It sounds like the cache bug again causing visual artifacts.

> I get what you’re going for with ClipboardManager, but as it is now the way some of the logic is split between Editor and ClipboardManager is a little suboptimal. Especially for the paste part, where the ClipboardManager essentially has the knowledge of when Editor is able to perform a paste. I’d prefer if the separation was a bit cleaner there, for example by managing only the clipboard contents in ClipboardManager and keeping the copying/pasting entirely within Editor. But tbh, it doesn’t matter much where exactly you draw the line between ClipboardManager and Editor as long as it is a clean line.

I tried to move the copy operation out of Editor but I guess it's a bit suboptimal indeed, sure we can move the canCopy, canPaste stuff out and leave the manager to only handle the clipboard content.

> I like the “unified copy/paste” for canvas and timeline behaviour a little better than before, but as you mentioned it’s still not particularly elegant. I considered that maybe it would help to keep track of which selection (timeline vs canvas) was most recently modified and pick one according to that, but I’m worried that it might make the entire mechanism seem somewhat unpredictable. I’d like to hear what @Jose-Moreno thinks about this matter.

This is sorta what I tried in the beginning but as you said, the problem with tracking which part was last modified is that if the user did the copy some time ago and think they copied from the timeline and not the canvas, they could be left in a confused state, since we don't visually show them what they have stored in their clipboard.

CandyFace commented on core_lib/src/interface/editor.cpp:
```diff
@@ -508,108 +502,193 @@ void Editor::resetAutoSaveCounter()
+    mScribbleArea->updateFrame(mFrame);
+}
+
+bool Editor::canCopy() const
```
I personally think it makes sense because you're dependant on everything from editor anyway, so why not simply make it a method with no arguments and make it easier to use instead of of mEditor->canCopy(mEditor->getCurrentFrame, mEditor->layers()->currentLayer)?

CandyFace replied:
Hmm so you mean that it's given that we always copy and paste from the current frame and layer? I could accept that argument, then I'll remove the parameters.

Jose-Moreno replied:
> Hmm so you mean that it's given that we always copy and paste from the current frame and layer? I could accept that argument, then I'll remove the parameters.

@CandyFace In the sense of working from and to an active layer, pasting a sequence of frames will occur from the initial frame we set the playhead at, however shouldn't copying / cutting be allowed to have the playhead positioned at any frame? I mean what should matter is the selection itself not the current frame position in that case 🤔 

The simplest example use case I can think of is I could select a non-juxtaposed sequence of frames to cut (i.e leave the even frame drawings, select the odd frames), and then scrub through them to see If I missed any so the playhead would be at the middle or the end of the selection. Sometimes, for efficiency, I could also move the playhead to the place I want to paste, but since the frames are selected I press CTRL + C assuming the selection takes priority over the playhead position.

What do you think?

CandyFace replied:
When you move the playhead position, you evidentially also change the current frame, so pasting will always occurs on the frame you stand on, not where you copied. The position of the playhead has no role in this, except when you paste the frames, which happens on the current frame.

Copying and pasting across multiple layers is still not a thing, and will not become a thing in this PR, so I think it's fine to say that the copy/paste logic assumes that we always make changes to the current layer.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+
+bool ClipboardManager::canCopyBitmapImage(BitmapImage* bitmapImage) const
+{
+    return bitmapImage != nullptr && bitmapImage->isLoaded() && !bitmapImage->bounds().isEmpty();
```
What would you suggest we do instead? It's different because as you say, checking a layer which is null shouldn't be possible where as a keyframe can indeed be null because that is the way we check against non existent frames.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+void ClipboardManager::copyBitmapImage(BitmapImage* bitmapImage, QRectF selectionRect)
+{
+    resetStates();
+    if (bitmapImage == nullptr || !bitmapImage->isLoaded()) { return; }
```
Right but it shouldn't crash to try either way, that's why the nullptr check is there. If the bitmap image is not loaded either, then something is very wrong and loading might cause other unwanted behaviors we don't want.

I'll change it to an assert instead.

Jose-Moreno commented:
@J5lx @CandyFace 
> I like the “unified copy/paste” for canvas and timeline behaviour a little better than before, but as you mentioned it’s still not particularly elegant. I considered that maybe it would help to keep track of which selection (timeline vs canvas) was most recently modified and pick one according to that, but I’m worried that it might make the entire mechanism seem somewhat unpredictable. I’d like to hear what @Jose-Moreno thinks about this matter.

I've just tested the current branch modifications, I personally feel it's good to have a simplistic behavior so the current "unified" clipboard works fine, that is, _If you copy a drawing, you paste a drawing, if you copy a frame, you paste a frame._ when I wanted to paste a frame specifically, I kept clicking on the empty frame where I wanted it to be so the playback cursor moves over the position and the layer gets selected.

In apps like Adobe Animate, there is a clear distinction between copying artwork vs copying frames and even layers on the timeline are treated differently, which is honestly a chore sometimes. 

Basically using CTRL + C / CTRL + V will ONLY copy/paste artwork on canvas and nothing more, while using CTRL + ALT + C / CTRL + ALT + V will copy/paste frames, but for the latter to work you NEED to select the actual frame exposure / empty frame at the desired frame dial position, and only then you are allowed to paste the frame. 

Copying and pasting layers however has to be done with a contextual menu, while frames used to be like this too, you can also assign shortcuts, but they are treated and stored as separate entities in the end and copying a layer is not the same as copying a frame nor a piece of vector drawing, and that piece of info will be kept there forever until you close the file or choose a different one for each.

Considering all of that Pencil2D right now feels a bit more flexible and free to maneuver for such tasks and since you are not meant to keep track in your head of what is copied and what isn't I feel it's great to make it unified.

Even if users remotely expect timeline "objects" to work different from canvas "objects" in the clipboard. If they are clearly shown that everything works with a single workflow (whatever is copied first, is pasted first) I don't think they will complain at all, since no matter what copying/pasting is an active behavior where you have to do the task sequentially while paying attention only to the element you want to copy and paste on that specific moment, and that should be good enough I think 🤔

Jose-Moreno commented:
On a separate topic I have a question. When I was pasting drawings in this branch, all the drawings kept getting pasted at the center of the canvas / camera at (0,0) instead of getting pasted on the same position where they were copied / cut from as usual (which is honestly a drawback) can either of you confirm If it's a bug with this branch? or is it a new behavior being implemented here? This doesn't happen in master (and honestly it shouldn't happen unless there's proper user input)

CandyFace commented:
> When I was pasting drawings in this branch, all the drawings kept getting pasted at the center of the canvas / camera at (0,0) instead of getting pasted on the same position where they were copied / cut from as usual (which is honestly a drawback) can either of you confirm If it's a bug with this branch?

That's a regression, i'll look into that.

CandyFace commented:
Alright after a lot more work being put into the next iteration, I think it's ready for review again.

Having used this build for some time, I've tweaked the layout some more, adjusted shortcuts to what I think works and modified the increase/decrease exposure behavior to something I personally think is more useful.

What's new and changed:
- Setting exposure is always done on a selection now.
- Timeline selection behaviors has been moved into its own subcategory under animation, see below.
<img width="429" alt="image" src="https://user-images.githubusercontent.com/1045397/107679160-8232d700-6c9c-11eb-9645-d43587fa5d05.png">

- Scrubber enhancement: will now be shown as hollow when hovering over.
- Fixed frames would be selected when when changing layer via shortcuts.
- Tests has been written to hopefully ensure that most of the new frame exposure functionality works as expected as well as some of the other changes I've implemented.

I'm still not sure about the actions in the menu but i think that maybe we can look into that in a separate PR.. I've been thinking that a lot of the actions in the menu should be revisited again, so we can make the most out of the menu categories.

Jose-Moreno commented:
@CandyFace I tested this recently and so far I saw everything working properly except for the minor issue with the decrease exposure command which you already fixed. 

In regards to 4v0v's initial PR I think everything has been addressed, and regarding the pending features from David's original proposal (moving frames + looping frames) those should be addressed in separate issues due to extra pre-requisites that haven't been implemented yet into Pencil2D (e.g resource referencing, frame instancing, multi-layer frame selection & manipulation, etc).

So far the copy, cut and remove frames features are working as expected and didn't seem to have any issue with the minor tests I ran. Though i'm a bit worried since after using these capabilities it became obvious that we really need the undo & redo system improvements to come soon to finalize the overall timeline manipulation UX.

Anyway, I'll dedicate some extra time trying to test the limits of this PR, but in regards to the original scope set in the description I believe it's well beyond what was proposed and it makes me extremely happy to finally see the result, congrats! 😄

Jose-Moreno commented:
@CandyFace Hi 👋 I wanted to ask something. I was testing this PR again and I noticed that the previous behaviour to add or subtract exposure has changed a bit. Before IIRC I could affect a single frame exposure depending on the position of the timeline cursor, however now it seems I have to specifically select the frame before being able to change the exposure.

I'm wondering if it's a restriction required for the past behaviour change we discussed (that the frames would be pushed in a certain way) or is it something else?

Having to select single frames seem to drop the efficiency while working as well as forcing the user to navigate the timeline just to use the actual shortcuts. 

For selecting a range of frame to change the exposure I don't see a problem, but I feel it is otherwise inefficient to select a single frame each time you want to change the exposure considering you might be occupied drawing most of the time. Additionally this kind of feature is always meant as a shorthand to avoid having to drag the frames themselves 🤔 

That said the current behaviour might not be as uncomfortable if there was a shortcut to quickly select and move the active layer frame selection between and across keyframes, but if possible, for now, I'd like to request your consideration to allow the single frame exposure modification to occur without requiring a selection, though multi-frame exposure can be kept as it is now since it makes sense.

What do you think? would that be possible or do you have a different concept you're trying to implement? If so I'd like to understand the idea behind it. 

Other than that it's as I mentioned before, everything else feels & works great. Maybe attending their respective undo/redo actions would be left. Thanks again for all the hard work 🙇

CandyFace commented:
@Jose-Moreno sorry for the very late reply... I was busy with other PR's so I haven't gotten my act together with this one..

A fair observation, more so because we have shortcuts for it, so I agree that it makes sense to be able to do so without messing with the timeline.
I have now re-implemented the logic required to make that work.

Also PR is now up to date, so @pencil2d/developers feel free to review 🙏

J5lx reviewed (changes requested):
I’ve let this PR sit for way too long again, haven’t I? Well, I’m sorry for that. Anyway, it’s actually shaping up quite nicely and I daresay we’re nearly there. First of all, here’s the usual minor stuff PR: CandyFace#17

And the remaining bugs:
- [x] It is not possible to change the exposure of a selection when there’s no keyframe on or before the current frame
- [x] When both a canvas selection and a frame selection exist, cutting copies the canvas selection but deletes the selected frames
- [x] The hover effect of the timeline scrub lingers even when the cursor is no longer hovering over it if the cursor “leaves” the scrub to the top or bottom

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,104 @@
+    resetStates();
+    if (vectorImage == nullptr || !vectorImage->isValid()) { return; }
+
+    // FIXME: handle vector selections, ie. independent strokes...
```
It's a fix me for the future, for when vector becomes prioritized again. It's not meant to be fixed in this PR

CandyFace commented on core_lib/src/managers/clipboardmanager.h:
```diff
@@ -0,0 +1,89 @@
+    std::map<int, KeyFrame*> getClipboardFrames() { return mFrames; }
+
+    bool framesTypeChanged(const Layer* layer) const { return layer->type() != mFramesType; }
+    bool anyFramesInClipbord() const { return !mFrames.empty(); }
```
Hmm you're right, I will make that change.

CandyFace commented on app/src/mainwindow2.cpp:
```diff
@@ -1391,16 +1406,22 @@ void MainWindow2::makeConnections(Editor* pEditor, TimeLine* pTimeline)
     connect(mTimeLine, &TimeLine::playButtonTriggered, mCommands, &ActionCommands::PlayStop);
 
+    // Clipboard state handling
+    connect(QApplication::clipboard(), &QClipboard::dataChanged, this, &MainWindow2::fetchClipboard);
```
the connect or the output of it, ie. the function?

the former is because QApplication should imo. not be a dependency for editor, since it's GUI related. I could make a signal to editor with the the clipboard as argument but otherwise I'd like to keep it where the GUI is represented.

J5lx replied:
Hmm. Fair enough.

CandyFace commented on core_lib/src/graphics/vector/vectorimage.h:
```diff
@@ -80,6 +80,8 @@ class VectorImage : public KeyFrame
     void deleteSelectedPoints();
     void removeVertex(int curve, int vertex);
 
+    bool isValid() const { return !mCurves.isEmpty(); }
```
Probably not and even if that was the case, we should probably also check for vertices.. I will change it to isEmpty then

CandyFace commented on core_lib/src/interface/timelinecells.cpp:
```diff
@@ -1068,9 +1129,22 @@ void TimeLineCells::mouseDoubleClickEvent(QMouseEvent* event)
-        editLayerProperties(layer);
+        if (mType == TIMELINE_CELL_TYPE::Tracks && (layerNumber != -1) && (frameNumber > 0) && layerNumber < mEditor->object()->getLayerCount())
+        {
+            layer->setFrameSelected(frameNumber, false);
```
Hmm honestly I don't recall why I did that but I would suspect it to be something with the state not being correct at the given time... maybe caused by another bug.

I will double check to make sure it's needed and otherwise remove it.

CandyFace commented:
> It is not possible to change the exposure of a selection when there’s no keyframe on or before the current frame
Could you elaborate what you mean by this ?

If you only have one keyframe, then you won't be able to add or remove exposure, because you will affect frames after that frame. 
If you select two frames on pos 2-3 and your current pos is on 5, then frame 3 and 5 will be exposed further.

>  The hover effect of the timeline scrub lingers even when the cursor is no longer hovering over it if the cursor “leaves” the scrub to the top or bottom

This seems like a minor issue, that I don't think is required to fix before merging. It doesn't break anything and as soon as the mouse enters the widget again, the hovering effect will disappear. 

> When both a canvas selection and a frame selection exist, cutting copies the canvas selection but deletes the selected frames
I will look into that.

J5lx commented:
> Can you show me an example?

What I mean is something like this:

![Screenshot_20210922_150858](https://user-images.githubusercontent.com/2063777/134349791-2c74a33b-eca9-4bdc-aed8-28eb0bb833ad.png)

Now if you are on frame 12, you can increase the exposure of the selected frames as expected. If you are on frame 2 on the other hand, you can’t. However, if you add a new keyframe at position 1 and then try to increase the exposure of the selected frames while you are still on position 2, it does work. Do you understand what I mean?

> This seems like a minor issue, that I don't think is required to fix before merging

In principle I agree, but I feel like this is the kind of thing that will never be fixed unless we fix it now. Tbh I don’t even think I would block the merge over this, but I’d still prefer for this to be fixed beforehand.

CandyFace commented on app/src/actioncommands.cpp:
```diff
@@ -600,6 +600,93 @@ Status ActionCommands::addNewKey()
+        currentLayer->removeKeyFrame(pos);
+    }
+
+    currentLayer->deselectAll();
```
Indeed it looks like a leftover.. will be removed.

CandyFace commented on core_lib/src/interface/editor.cpp:
```diff
@@ -512,107 +508,194 @@ void Editor::resetAutoSaveCounter()
-        VectorImage *vectorImage = static_cast<VectorImage*>(layer->getLastKeyFrameAtPosition(currentFrame()));
-        if (vectorImage == nullptr) { return; }
-        g_clipboardVectorImage = *vectorImage;  // copy the image
+        currentLayer->deselectAll();
```
Indeed, not sure why I didn't consider doing that when I initially implemented it. Will change.

CandyFace commented on core_lib/src/interface/editor.cpp:
```diff
@@ -512,107 +508,194 @@ void Editor::resetAutoSaveCounter()
+    bool canPasteState = canPaste(layer);
+
+    emit canCopyChanged(canCopyState);
+    emit canCutChanged(canCopyState);
```
You're right, there's no situation where you can't cut but can copy, at least not in pencil2d. I will make the change.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,104 @@
+    return true;
+}
+
+void ClipboardManager::setFromSystemClipboard(const QClipboard *clipboard, const Layer* layer)
```
Because I had an idea that clipboard wouldn't always have newest state, but that does not seem to be the case. I will remove the argument.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,104 @@
+}
+
+bool ClipboardManager::updateIfNeeded(const Layer* layer)
+{
```
It was yes, because I needed the last stored position... however looking at the solution again, that was a bad idea. It's much cleaner to get the pos from outside. Fixed in latest commit.

CandyFace commented on core_lib/src/managers/clipboardmanager.cpp:
```diff
@@ -0,0 +1,104 @@
+
+    mBitmapImage = BitmapImage(mLastBitmapPosition, clipboardImage);
+    return true;
+}
```
Fixed in latest commit. Since we didn't use the return value, I decided to get rid of it and use setFromSystemClipboard instead.

CandyFace commented:
I believe I've taken care of all comments and reported bugs now.

J5lx reviewed (changes requested):
Alright, I fixed some small oversights with one of the changes, other than that we’re pretty much there. There’s just one small bug left. In your most recent commit, vector pasting broke because of the introduction of resetStates() in setFromSystemClipboard(). Turns out that this is in fact a problem because setFromSystemClipboard(), and thus resetStates(), is called unconditionally in Editor::paste(), erasing the vector clipboard contents. Bitmap does not have that issue because anything that is copied from a bitmap layer is also written to the system clipboard, and later read back from the system clipboard by setFromSystemClipboard(). Still, doesn’t seem quite as intended. Personally I’m wondering whether it’s even necessary for setFromSystemClipboard() to be called in paste() since we already have clipboardChanged(). Any thoughts?

J5lx commented:
Oh and btw, would you mind if I squash-merged this PR due to the huge number of commits and number of iterations that this PR has seen?

CandyFace commented:
> fixed some small oversights with one of the changes, other than that we’re pretty much there
👍 yup that was indeed an oversight.. I managed to turn it around wrong, niice. 🤦 

> Turns out that this is in fact a problem because setFromSystemClipboard(), and thus resetStates(), is called unconditionally in Editor::paste(), erasing the vector clipboard contents

Ah... you're right, welp I knew there was a reason why I did not apply it specifically in that case, other than because I needed the last position too... that's what happens when you leave a PR hanging for too long, you forget the implementation and intend. I've fixed it now and added a comment.

> Personally I’m wondering whether it’s even necessary for setFromSystemClipboard() to be called in paste() since we already have clipboardChanged(). Any thoughts?

It is necessary, at least on mac because QClipboard::dataChanged doesn't catch the change if the application is on the foreground while you copy to clipboard, eg. by taking a screenshot.

J5lx reviewed (approved):
Alright, awesome! Let’s finally merge this!

J5lx commented:
Seems that GitHub didn’t recognize my merge commit so I’m closing this manually. But rest assured that everything has been merged!

Jose-Moreno commented:
@CandyFace Hey first of all nice job and thank you for all the hard work!!! I'm testing this after the automated build finished so far so good 😄 

I wanted to ask about one thing that captured my attention. I thought you had implemented a way to freeze the frames in place so when someone moved a frame it wouldn't shift it over to the side. Was that removed from this PR? or is it in a different one?

J5lx commented:
That’s #1495 :)
